src/Sets/Ball1.jl

```diff
@@ -179,19 +181,23 @@ julia> [.5, 1.5] ∈ B
 true
 ```
 """
-function ∈(x::AbstractVector{N}, B::Ball1{N})::Bool where {N<:Real}
-    @assert length(x) == dim(B)
+function ∈(x::AbstractVector{N}, B::Ball1{N}, failfast::Bool=false) where {N<:Real}
+    @assert length(x) == dim(B) "a $(length(x))-dimensional vector is " *
+        "incompatible with a $(dim(B))-dimensional set"
     sum = zero(N)
-    for i in eachindex(x)
-        sum += abs(B.center[i] - x[i])
+    for (i, xi) in enumerate(x)
+        sum += abs(B.center[i] - xi)
+        if failfast && sum > B.radius
```

Comment:
Here, `failfast` is a loop invariant and can be extracted out of the loop. I'm curious if Julia automatically detects this.

Reply:
I always thought that `Bool` arguments are compiled away.

I tried an example and first was shocked: the `goto #8 if not failfast` instruction below looks like this is not the case.
But then I searched a bit and found [this](https://discourse.julialang.org/t/how-does-constant-propagation-work/22735). So I am still quite convinced that it works, but I do not see a quick way to prove it.

```julia
julia> @code_lowered ∈([1.], Ball1([1.], 1.), false)
CodeInfo(
1 ──       Core.NewvarNode(:(sum))
│          Core.NewvarNode(:(@_6))
│    %3  = Base.getproperty(LazySets, :are_assertions_enabled)
│    %4  = (%3)()
└───       goto #5 if not %4
2 ── %6  = LazySets.length(x)
│    %7  = LazySets.dim(B)
│    %8  = %6 == %7
└───       goto #4 if not %8
3 ──       goto #5
4 ── %11 = Base.getproperty(Base.Main, :Base)
│    %12 = Base.getproperty(%11, :string)
│    %13 = LazySets.length(x)
│    %14 = Base.string("a ", %13, "-dimensional vector is ")
│    %15 = LazySets.dim(B)
│    %16 = Base.string("incompatible with a ", %15, "-dimensional set")
│    %17 = %14 * %16
│    %18 = (%12)(%17)
│    %19 = Base.AssertionError(%18)
└───       Base.throw(%19)
5 ┄─       sum = LazySets.zero($(Expr(:static_parameter, 1)))
│    %22 = LazySets.enumerate(x)
│          @_6 = Base.iterate(%22)
│    %24 = @_6 === nothing
│    %25 = Base.not_int(%24)
└───       goto #13 if not %25
6 ┄─ %27 = @_6
│    %28 = Core.getfield(%27, 1)
│    %29 = Base.indexed_iterate(%28, 1)
│          i = Core.getfield(%29, 1)
│          @_9 = Core.getfield(%29, 2)
│    %32 = Base.indexed_iterate(%28, 2, @_9)
│          xi = Core.getfield(%32, 1)
│    %34 = Core.getfield(%27, 2)
│    %35 = sum
│    %36 = Base.getproperty(B, :center)
│    %37 = Base.getindex(%36, i)
│    %38 = %37 - xi
│    %39 = LazySets.abs(%38)
│          sum = %35 + %39
└───       goto #8 if not failfast
7 ── %42 = sum
│    %43 = Base.getproperty(B, :radius)
│          @_10 = %42 > %43
└───       goto #9
8 ──       @_10 = false
9 ┄─       goto #11 if not @_10
10 ─       return false
11 ─       @_6 = Base.iterate(%22, %34)
│    %50 = @_6 === nothing
│    %51 = Base.not_int(%50)
└───       goto #13 if not %51
12 ─       goto #6
13 ┄ %54 = sum
│    %55 = Base.getproperty(B, :radius)
│    %56 = %54 <= %55
└───       return %56
)
```